### PR TITLE
COUUnit able to evaluate its current OpenStackRelease

### PR DESCRIPTION
**Pull request by gabrielcocenza**
- move the responsability to evaluate OpenStackRelease from units to the COUUnit instead of the Application
- adapt unit tests

### CODE REVIEW

**Comment by gabrielcocenza** [comment]
With those changes it will be easier to not include units of `nova-compute` and colocated OpenStack apps that already got upgraded for a certain target

File: cou/apps/channel_based.py [hunk 1]
```diff
@@ -28,17 +29,17 @@
 class ChannelBasedApplication(OpenStackApplication):
     """Application for charms that are channel based."""
 
-    def _get_latest_os_version(self, unit: COUUnit) -> OpenStackRelease:
-        """Get the latest compatible OpenStack release based on the channel.
+    @property
```
Comment (rgildein) [inline cou/apps/channel_based.py:32]: Why we need this here? It's not same version as we have in base.py?

**Review by rgildein — COMMENTED** [review]
I do not see the benefit from this approach, can you explain it more?
Like this we are mixing logic of OpenStackApplication and COUApplication, let say we have model with some apps (e.g. telegraf), which are not related with OpenStack anyhow, and we do `app.os_release`. We will get an error saying that there is no compatible version, which is confusing. I know we should never run `os_release` on not OpenStackApplication.

File: cou/apps/channel_based.py (hunk 1 above)
Comment (gabrielcocenza) [reply cou/apps/channel_based.py:32]: > It's not same version as we have in base.py?
No. Channel base apps doesn't have a workload version. We decide the OpenStackRelease by what the charm channel says

Comment (rgildein) [reply cou/apps/channel_based.py:32]: I see, `os_versions[unit.os_release].append(unit.name)` and `os_versions[self.channel_codename].append(unit.name)`.

**Comment by gabrielcocenza** [comment]
> Like this we are mixing logic of OpenStackApplication and COUApplication, let say we have model with some apps (e.g. telegraf), which are not related with OpenStack anyhow, and we do app.os_release. We will get an error saying that there is no compatible version, which is confusing. I know we should never run os_release on not OpenStackApplication.

Good point, I didn't think about this issue of mixing. What I'm currently missing is to easily access the `OpenStackRelease` just having the Units objects and not the Application itself. Maybe we can "overwrite"  the units in the factory once we know that the application is supported.

**Comment by rgildein** [comment]
I do not think that this will be needed, so I'm closing it. Feel free to re-open it, if you come with another approach.

**Pull request closed without merging**